src/services/mapCode.ts

```diff
@@ -0,0 +1,365 @@
+import {
+    Block,
+    ClassElement,
+    ClassLikeDeclaration,
+    createSourceFile,
+    factory,
+    FileTextChanges,
+    forEach,
+    forEachChild,
+    formatting,
+    hasJSDocNodes,
+    InterfaceDeclaration,
+    isBlock,
+    isClassElement,
+    isClassLike,
+    isNamedDeclaration,
+    isSourceFile,
+    isSourceFileJS,
+    isTypeElement,
+    LanguageServiceHost,
+    NamedDeclaration,
+    Node,
+    NodeArray,
+    Program,
+    ScriptTarget,
+    SourceFile,
+    SourceMapper,
+    Statement,
+    SyntaxKind,
+    sysLog,
+    textChanges,
+    TextSpan,
+    TypeElement,
+    UserPreferences,
+} from "./_namespaces/ts";
+import {
+    ChangeTracker,
+} from "./textChanges";
+
+/** @internal */
+export function mapCode(
+    mappings: { sourceFile?: SourceFile; contents: string[]; focusLocations?: TextSpan[][]; }[],
+    _program: Program,
+    host: LanguageServiceHost,
+    formatContext: formatting.FormatContext,
+    preferences: UserPreferences,
+    _sourceMapper: SourceMapper,
+    _updates?: FileTextChanges[],
+): FileTextChanges[] {
+    try {
+        return textChanges.ChangeTracker.with(
+            { host, formatContext, preferences },
+            changeTracker => {
+                mappings.forEach(({ sourceFile, contents, focusLocations }) => {
+                    if (!sourceFile) {
+                        // TODO(@zkat): uhhh... do something about file-less mappings. Not supported for now.

```

Comment:
If this is still an unknown, why do we want to still support an undefined `file` in `MapCodeRequestDocumentMapping`?

Reply:
It's part of the protocol that VS/VSCode send already, but since we're doing fuzzy heuristics, I think it's ok for us to accept it in the protocol, but ignore it for now.